### PR TITLE
feat(providers): load previous logs before subscribing

### LINKED ISSUE #988: Stream events, including past ones
**Is your feature request related to a problem? Please describe.**

Currently there are two ways to query events:
- with a filter for a given range:`contract.events().from_block(block_number).query().await`. This allows to query events that have been emitted in the past.
- streaming new events: `contract.events().stream().await`. This will only yield events that are emitted after the stream has been created.

**Describe the solution you'd like**
Ideally, there would be a way to get a stream of events from a given point in the past. I.e. a function that combines both above functions.

**Additional context**
See https://github.com/gakonst/ethers-rs/issues/986

### PR DESCRIPTION
**Pull request by meetmangukiya**
Load previous logs and stream it back to the user before establishing
a new stream for streaming future logs.

Closes #988

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [x] Updated the changelog

### CODE REVIEW

File: ethers-providers/src/provider.rs [hunk 1]
```diff
@@ -1102,9 +1102,29 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
...
+                } else {
+                    self.get_logs(filter).await
+                }
+            }
+            FilterBlockOption::AtBlockHash(_block_hash) => self.get_logs(filter).await,
+        };
+
+        if logs.is_err() {
+            return Err(logs.err().unwrap())
+        }
```
Comment (mattsse) [inline ethers-providers/src/provider.rs:1118]: can't we use the `?` here?

File: ethers-providers/src/pubsub.rs [hunk 2]
```diff
@@ -31,6 +31,8 @@ pub struct SubscriptionStream<'a, P: PubsubClient, R: DeserializeOwned> {
     /// The subscription's installed id on the ethereum node
     pub id: U256,
 
+    loaded_elements: Vec<R>,
```
Comment (mattsse) [inline ethers-providers/src/pubsub.rs:34]: looks like we can use a VecDeque here to easily `pop_front`?

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nits

File: ethers-providers/src/pubsub.rs (hunk 2 above)
Comment (meetmangukiya) [reply ethers-providers/src/pubsub.rs]: good catch

File: ethers-providers/src/provider.rs [hunk 3]
```diff
@@ -1102,9 +1104,24 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
     where
         P: PubsubClient,
     {
+        let loaded_logs = match filter.block_option {
+            FilterBlockOption::Range { from_block, to_block: _ } => {
+                if from_block.is_none() {
+                    vec![]
+                } else {
+                    self.get_logs(filter).await?
```
[the remark this answers is not in the document]
Comment (gakonst) [reply ethers-providers/src/provider.rs:1112]: There is an issue for this already I believe, #954 cc @meetmangukiya

**Pull request merged**